extension/apmconfigextension/elastic/centralconfig/fetcher.go

```diff
@@ -0,0 +1,97 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package centralconfig // import "github.com/elastic/opentelemetry-collector-components/extension/apmconfigextension/elastic/centralconfig"
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/elastic/opentelemetry-collector-components/extension/apmconfigextension/apmconfig"
+	"github.com/elastic/opentelemetry-lib/agentcfg"
+	"github.com/open-telemetry/opamp-go/protobufs"
+	semconv "go.opentelemetry.io/collector/semconv/v1.26.0"
+	"go.uber.org/zap"
+)
+
+const configContentType = "text/json"
+
+var _ apmconfig.RemoteConfigClient = (*fetcherAPMWatcher)(nil)
+
+type fetcherAPMWatcher struct {
+	configFetcher agentcfg.Fetcher
+	cacheDuration time.Duration
+
+	// OpAMP instanceID to service mapping
+	uidToService map[string]agentcfg.Service
+
+	logger *zap.Logger
+}
+
+func NewFetcherAPMWatcher(fetcher agentcfg.Fetcher, cacheDuration time.Duration, logger *zap.Logger) *fetcherAPMWatcher {
+	return &fetcherAPMWatcher{
+		configFetcher: fetcher,
+		cacheDuration: cacheDuration,
+		uidToService:  make(map[string]agentcfg.Service),
```

Comment:
With every call of `RemoteConfig()` this map `uidToService` is populated and never cleaned up. Therefore it just grows without limits.There is also the risk, that `RemoveConfig()` fetches something from this map - via `serviceParams, ok := fw.uidToService[string(agentMsg.GetInstanceUid())]` that is out of date and not work any longer.

Reply:
Good point, I think removals should be done on the agent disconnects (AgentDisconnect message), but it would be more appropriate to handle in a follow-up PR https://github.com/elastic/opentelemetry-collector-components/issues/398